### PR TITLE
Automatic redshift for observational data

### PR DESCRIPTION
**Pull request by EvgeniiChaikin**
**1. Why this PR**
Since in the newer version of the pipeline redshift bracketing was not properly working, and since there has been increasing demand to have automatically selected redshifts for observational data, I have made this PR.

**2. Usage:**

- One extra optional parameter in yaml files in the observational data section: **automatic_redshift** = [`true` or `false`].
- Default value (if the parameter is not present in the yalm file): `false` 
- If `true`, the name of the observational data must be provided excluding the `_z???p???.hdf5` part. Regual-expression search assumes that the files have the names `{filename}_z???p???.hdf5`, where `?` is an integer number from 0 through 9.

**3. Example of working .yaml file**

```
stellar_mass_halo_mass_MBN98_centrals_100:
  type: "scatter"
  select_structure_type: 10
  comment: "Centrals only"
  legend_loc: "upper left"
  x:
    quantity: "masses.mass_bn98"
    units: Solar_Mass
    start: 1e8
    end: 1e13
  y:
    quantity: "apertures.mass_star_100_kpc"
    units: Solar_Mass
    start: 9e3
    end: 1e12
  median:
    plot: true
    log: true
    adaptive: true
    number_of_bins: 25
    start:
      value: 1e8
      units: Solar_Mass
    end:
      value: 1e13
      units: Solar_Mass
  metadata:
    title: Stellar Mass-Halo Mass relation (100 kpc aperture, $M_{\rm BN98}$, centrals only)
    caption: Includes only central haloes.
    section: Stellar Mass-Halo Mass
  observational_data:
    - filename: GalaxyStellarMassHaloMass/Behroozi2019
      automatic_redshift: true
    - filename: GalaxyStellarMassHaloMass/Schaye2015_Ref_100_z0p0.hdf5
```

**4. Special cases**

- If **automatic_redshift** is `true` and no filename matches were found (likely because the matching pattern was chosen incorrectly), then the following exception is raised:
```
Exception: No matches for the regular-expression search pattern 'GalaxyStellarMassHaloMass/Behroozi2019_z\d{3}p\d{3}.hdf5' were found. Choose a correct search pattern or set automatic redshift parameter to false
```
- If **automatic_redshift** is set to anything but `true` or `false` ( `=`  not present in the file), then the following error is raised (`automatic_redshift: 0`):
` TypeError: Automatic redshift must be either true or false, not 0`

**5. Results**

Without changing .yaml files, velociraptor, or comparison data, I feed the velociraptor library with three SWIFT snapshots at `z={0, 2, 4}`. The results are

z_0_stellar_mass_halo_mass_MBN98_centrals_100
z2_stellar_mass_halo_mass_MBN98_centrals_100
z4_stellar_mass_halo_mass_MBN98_centrals_100

As can be seen, Berhoozi2019 data' redshift follows the redshift of the snapshot (because **automatic_redshift**: `true`) while EAGLE data is always at `z=0` (because **automatic_redshift** was not provided and is thus equal to `false`)

@JBorrow this PR is just an example of how in my opinion automatic redshift should work. If you think there is a better way of doing that (there is always a better way), I will be happy to discuss it with you and apply the necessary changes to this PR or close this PR.

### CODE REVIEW

**Comment by JBorrow** [comment]
Looks like a good idea! Do we not want to make sure to keep the manual bracketing, though? For datasets that do not have multiple redshifts?

**Comment by JBorrow** [comment]
Also, I (in general) don't like to do things based on filenames. Maybe we should modify the observational data files to allow for multiple redshifts?

**Comment by EvgeniiChaikin** [comment]
> Looks like a good idea! Do we not want to make sure to keep the manual bracketing, though? For datasets that do not have multiple redshifts?

Could you give me an example of where the manual bracketing may be useful? 

For the case of observation data available only at a single redshift, the code in this PR will just go ahead and plot the data from that single file with the observational data. I do not see how this will change if we use the redshift bracketing.

**Comment by EvgeniiChaikin** [comment]
> Also, I (in general) don't like to do things based on filenames. Maybe we should modify the observational data files to allow for multiple redshifts?

That'd be better. I agree. We will then need to think of how to arrange data at multiple z inside a single .hdf5 file, and also what going to be stored in **assosiate_redshift**. A list of floats instead of a single float?

**Comment by EvgeniiChaikin** [comment]
@JBorrow  Here is the structure of an hdf5 file for typical single-`z` data. There are **512** entries in the y array in the data file.

```
Behroozi2019_z000p000.hdf5
├cosmology
│ └9 attributes:
│   ├H0: 67.74
│   ├Neff: 3.046
│   ├Ob0: 0.0486
│   ├Ode0: 0.6910098821161554
│   ├Om0: 0.3075
│   ├Tcmb0: 2.7255
│   ├m_nu: [0.   0.   0.06]
│   ├m_nu_units: 'eV'
│   └name: 'Planck15'
├metadata
│ └6 attributes:
│   ├bibcode: '2019MNRAS.488.3143B'
│   ├citation: 'Behroozi et al. (2019)'
│   ├comment: 'The data is taken f...function halo mass.'
│   ├name: 'Fit to the stellar ...s relation at z=0.0'
│   ├plot_as: 'line'
│   └redshift: 0.0
├x
│ ├2 attributes:
│ │ ├comoving: True
│ │ └description: 'Halo Mass ($M_{\\rm BN98}$)'
│ └values [float64: 512]
│   └2 attributes:
│     ├unit_registry: void(b'\x80\x03\x7D\x71\x00\x2E')
│     └units: 'Msun'
└y
  ├2 attributes:
  │ ├comoving: True
  │ └description: 'Galaxy Stellar Mass'
  └values [float64: 512]
    └2 attributes:
      ├unit_registry: void(b'\x80\x03\x7D\x71\x00\x2E')
      └units: 'Msun'

```

Now suppose we have such data **at N different redshifts** and we want to combine these into a single .hdf5 file. In this case, I suggest we 

- store an **N**-length array of redshifts into **metadata.redshift**. 
- In addition to that, we create a new attribute, **metadata.number_of_redshifts** (perhaps we could choose a better name?), which will be equal to **N**
- **y.values** will become an **Nx512** matrix instead of a 512-length 1D array.

I will invite @MatthieuSchaller to join this discussion because he's been recently making EAGLE data available at multiple z.

**Comment by EvgeniiChaikin** [comment]
> Also, I (in general) don't like to do things based on filenames. Maybe we should modify the observational data files to allow for multiple redshifts?

@JBorrow Should be working now. Please check out [THIS NEW PR](https://github.com/SWIFTSIM/velociraptor-comparison-data/pull/51) in the comparison data repository to see how a multi-z .hdf5 file is structured.

Note that if the file contains observational data at multiple z, then the data with z closest to the snapshot's z is found **by default**

File: velociraptor/observations/objects.py [hunk 1]
```diff
@@ -569,16 +576,21 @@ def plot_on_axes(self, axes: Axes, errorbar_kwargs: Union[dict, None] = None):
...
+        # Check whether the data is available at several redshifts
+        if line_idx >= 0:
+            observational_data_y = self.y[line_idx, :]
+            data_label = f"{self.citation} ($z={self.redshift[line_idx]:.1f}$)"
+        else:
+            observational_data_y = self.y
+            data_label = f"{self.citation} ($z={self.redshift:.1f}$)"
 
         axes.errorbar(
             self.x,
```
Comment (JBorrow) [inline velociraptor/observations/objects.py:588]: x may not be the same for all redshifts

File: velociraptor/observations/objects.py [hunk 2]
```diff
@@ -531,6 +532,12 @@ def plot_on_axes(self, axes: Axes, errorbar_kwargs: Union[dict, None] = None):
 
         errorbar_kwargs: dict
             Optional keyword arguments to pass to plt.errorbar.
+
+        line_idx: int
```
Comment (JBorrow) [inline velociraptor/observations/objects.py:536]: I don't like leaving users to specify which index... perhaps we can have them give a redshift and the closest data is returned?

**Comment by JBorrow** [comment]
Don't think that we can use multidimensional arrays for the multi-z data. They may not be the same length.

File: velociraptor/observations/objects.py (hunk 2 above)
Comment (EvgeniiChaikin) [reply velociraptor/observations/objects.py:536]: In this version, the user never gets to choose the index. Instead, it is done automatically via minimising **abs**(observational data z - snapshot z)

File: velociraptor/observations/objects.py (hunk 1 above)
Comment (EvgeniiChaikin) [reply velociraptor/observations/objects.py:588]: Well, yes you are right.

I will generalise it then to include the case wherein x are different for different z.

File: velociraptor/observations/objects.py (hunk 2 above)
Comment (EvgeniiChaikin) [reply velociraptor/observations/objects.py:536]: Actually, yes, indeed! There is a much better way of doing this.

File: velociraptor/observations/objects.py (hunk 1 above)
Comment (JBorrow) [reply velociraptor/observations/objects.py:588]: Note the comment below about x and y are different _lengths_ for different z. I think we need to be much more careful about this before forging ahead with an implementation as we'll need to change quite a lot of files.

**Comment by EvgeniiChaikin** [comment]
@JBorrow Is it better now? I can bring back the z bracketing feature if you want. Although I still cannot imagine a situation where that could be useful.

**Comment by EvgeniiChaikin** [comment]
This PR is closed because there is another PR with the same feature

Discussion continues [HERE](https://github.com/SWIFTSIM/velociraptor-python/pull/38)

**Pull request closed without merging**